ocaml/xapi/vm_evacuation.ml

```diff
@@ -105,7 +105,19 @@ let ensure_no_vms ~__context ~rpc ~session_id ~evacuate_timeout =
 
       evacuate ());
 
-  log_and_ignore_exn (fun () -> get_running_domains () |> shutdown)
+  log_and_ignore_exn (fun () -> get_running_domains () |> shutdown);
+
+  (* The driver domains shouldn't be in use at this point, so we should be able to
+   * shut them down *)
+  log_and_ignore_exn @@ fun () ->
+      get_resident_vms ~__context ~self:host |> fst
+      |> List.filter_map (fun vm ->
```

Comment:
This is mixing a functional map with a side effect inside the predicate. I think I would prefer to filter first and to iterate separately.

Reply:
I've split this so it first does the filtering for pbds, vms and then the unplugs and shutdowns.